mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+import itertools
+import random
+from collections.abc import Iterable
+from functools import cache, cached_property
+from typing import Callable, Optional
+
+from .. import Agent
+
+Coordinate = tuple[int, int]
+
+
+class Cell:
+    __slots__ = ["coordinate", "_connections", "agents", "capacity", "properties"]
+
+    def __init__(self, i: int, j: int, capacity: int | None = 1) -> None:
+        self.coordinate = (i, j)
+        self._connections: list[Cell] = []
+        self.agents: list[Agent] = []
+        self.capacity = capacity
+        self.properties: dict[str, object] = {}
+
+    def connect(self, other) -> None:
+        """Connects this cell to another cell."""
+        self._connections.append(other)
+
+    def disconnect(self, other) -> None:
+        """Disconnects this cell from another cell."""
+        self._connections.remove(other)
+
+    def add_agent(self, agent: Agent) -> None:
+        """Adds an agent to the cell."""
+        if self.capacity and len(self.agents) >= self.capacity:
+            raise Exception("ERROR: Cell is full")
+        if isinstance(agent.cell, Cell):
+            agent.cell.remove_agent(agent)
+        self.agents.append(agent)
+        agent.cell = self
+
+    def remove_agent(self, agent: Agent) -> None:
+        """Removes an agent from the cell."""
+        self.agents.remove(agent)
+        agent.cell = None
+
+    @property
+    def is_empty(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
+    def neighborhood(self, radius=1, include_center=False):
+        if radius == 0:
+            return {self: self.agents}
+        if radius == 1:
+            return CellCollection(
+                {neighbor: neighbor.agents for neighbor in self._connections}
+            )
+        else:
+            neighborhood = {}
+            for neighbor in self._connections:
+                neighborhood.update(neighbor.neighorhood(radius - 1, include_center))
+            if not include_center:
+                neighborhood.pop(self, None)
```

Comment:
Does this go right in the Grid? Because the centers aren't included there artwork. So shouldn't it be added if include_center is true, instead of popped if false?

Reply:
This is correct. The "center"cell enters the neighborhood indirectly from the neighboring cells (assuming we always have two-way connections between cells)